From f51e8e24094818bd5535b51b5d5388b3d6891805 Mon Sep 17 00:00:00 2001
From: David Riley <davidriley@chromium.org>
Date: Thu, 11 Oct 2018 15:50:09 -0700
Subject: [PATCH 10/61] CHROMIUM: drm/virtio: Add interfaces to share dma bufs
 via virtwl.

BUG=chromium:892342
TEST=xeyes repeatedly does not fail

Change-Id: Ib74f0992bd9ee2bde10097722c361cdd3e94b64f
Signed-off-by: David Riley <davidriley@chromium.org>
Reviewed-on: https://chromium-review.googlesource.com/1351815
Reviewed-by: Tomasz Figa <tfiga@chromium.org>
Reviewed-by: Zach Reizner <zachr@chromium.org>
Reviewed-by: Gurchetan Singh <gurchetansingh@chromium.org>
[rebase54(groeck):
    rebase53(rrangel):
        * Context conflict
        * Added missing includes.
]
Signed-off-by: Guenter Roeck <groeck@chromium.org>
---
 drivers/gpu/drm/virtio/virtgpu_object.c | 21 +++++++++++++++++++
 include/drm/virtio_drm.h                | 27 +++++++++++++++++++++++++
 2 files changed, 48 insertions(+)
 create mode 100644 include/drm/virtio_drm.h

diff --git a/drivers/gpu/drm/virtio/virtgpu_object.c b/drivers/gpu/drm/virtio/virtgpu_object.c
index 408a76029f09..29cf152ca61f 100644
--- a/drivers/gpu/drm/virtio/virtgpu_object.c
+++ b/drivers/gpu/drm/virtio/virtgpu_object.c
@@ -25,6 +25,7 @@
 
 #include <linux/dma-mapping.h>
 #include <linux/moduleparam.h>
+#include <linux/dma-buf.h>
 
 #include "virtgpu_drv.h"
 
@@ -261,3 +262,23 @@ int virtio_gpu_object_create(struct virtio_gpu_device *vgdev,
 	drm_gem_shmem_free_object(&shmem_obj->base);
 	return ret;
 }
+
+int virtio_gpu_dma_buf_to_handle(struct dma_buf *dma_buf, bool no_wait,
+				 uint32_t *handle)
+{
+	struct virtio_gpu_object *qobj;
+	struct virtio_gpu_device *vgdev;
+
+	if (dma_buf->ops != &virtgpu_dmabuf_ops)
+		return -EINVAL;
+
+	qobj = gem_to_virtio_gpu_obj(dma_buf->priv);
+	vgdev = (struct virtio_gpu_device *)qobj->base.base.dev->dev_private;
+	if (!qobj->create_callback_done && !no_wait)
+		wait_event(vgdev->resp_wq, qobj->create_callback_done);
+	if (!qobj->create_callback_done)
+		return -ETIMEDOUT;
+
+	*handle = qobj->hw_res_handle;
+	return 0;
+}
diff --git a/include/drm/virtio_drm.h b/include/drm/virtio_drm.h
new file mode 100644
index 000000000000..58e75777bf5b
--- /dev/null
+++ b/include/drm/virtio_drm.h
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Virtio GPU interfaces for sharing data.
+ *
+ * Copyright (C) 2018 Google, Inc.
+ */
+
+#ifndef _VIRTIO_DRM_H
+#define _VIRTIO_DRM_H
+
+#include <linux/dma-buf.h>
+#include <linux/types.h>
+
+/*
+ * The following methods are to share dma bufs with a host via the
+ * virtio Wayland (virtwl) device.
+ */
+
+/*
+ * Converts the given dma_buf to the virtio-gpu specific resource handle
+ * backing the dma_buf, waiting for creation to be confirmed by the host
+ * if necessary.
+ */
+extern int virtio_gpu_dma_buf_to_handle(struct dma_buf *buf, bool no_wait,
+					uint32_t *handle);
+
+#endif /* _VIRTIO_DRM_H */
-- 
2.25.1

